keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -697,43 +699,51 @@ generatedFrom:
   apiGroup: dra.example.com
   uid: foobar-uid
 
-vendorParameters:
-  # A vendor can put any kind of object here to pass the configuration
-  # parameters down to the kubelet plugin. In this case, the vendor
-  # driver simply copied the entire CR. It could also be some
-  # separate, smaller configuration type.
-  #
-  # Beware that ResourceClaimParameters have separate RBAC rules than
-  # the vendor CRD, so information included here may get visible
-  # to more users than the original CRD. Both objects are in the same
-  # namespace.
-  kind: CardParameters
-  apiVersion: dra.example.com/v1alpha1
-  metadata:
-    name: my-parameters
-    namespace: user-namespace
-    uid: foobar-uid
-  ...
-  spec:
-    minimumRuntimeVersion: v12.0.0
-    minimumMemory: 32Gi
-    sharing:
-      strategy: TimeSliced
-
-driverRequests:
-- driverName: cards.dra.example.com
-  requests:
-  # Each entry here is a request for one resource.
-  - namedresources:
+requests:
+- vendorParameters:
+    # A vendor can put any kind of object here to pass the configuration
+    # parameters down to the kubelet plugin. In this case, the vendor
+    # driver simply copied the entire CR. It could also be some
+    # separate, smaller configuration type.
+    #
+    # Beware that ResourceClaimParameters have separate RBAC rules than
+    # the vendor CRD, so information included here may get visible
+    # to more users than the original CRD. Both objects are in the same
+    # namespace.
+    kind: CardParameters
+    apiVersion: dra.example.com/v1alpha1
+    metadata:
+      name: my-parameters
+      namespace: user-namespace
+      uid: foobar-uid
+    ...
+    spec:
+      minimumRuntimeVersion: v12.0.0
+      minimumMemory: 32Gi
+      sharing:
+        strategy: TimeSliced
+  # Each entry here is a request for one resource. Entries could have
+  # their own vendor parameters (not shown here).
+  subRequests:
+  - namedResources:
```

Comment:
Before,  "requests" inside "driverRequests" made some sense because all "requests" were related to a single driver. Now the only remaining reason is that we need a place to put vendor parameters that are meant for more than one individual resource request.

@klueska: do you think that we can perhaps flatten this after all? Then a claim would have only "requests" where each request has vendor parameters and the selector.

Reply:
Sorry, I'm having a hard time paging this back in. What are you proposing exactly? Can you show me a before and after?

Reply:
Before is what this KEP has:
```
requests:
  - vendorParameters: <parameters for all sub-requests>
    subRequests:
    - vendorParameters: <parameters for device A>
      namedResources:
        <device A>
    - vendorParameters: <parameters for device B>
      namedResources:
        <device B>
```

If we were to drop one level, that would become:
```
requests:
- vendorParameters: <parameters for device A>
  namedResources:
     <device A>
- vendorParameters: <parameters for device B>
   namedResources:
     <device B>
```

Reply:
With one level removed, "common" parameters would need to be repeated for each device.

Reply:
Or set once and then applied to all.

Reply:
I think we can drop one level of indirection by moving the list of sub-requests into the model. When combined with renamed "named resources" model to "device" model (see [Slack](https://kubernetes.slack.com/archives/C0409NGC1TK/p1714904585320519?thread_ts=1714800983.574339&cid=C0409NGC1TK)), the result looks fairly natural to me.

Here I am also using a one-of-many for the parameters because instead of "vendor" setup parameters we might at some point also have "in-tree" setup parameters.

In addition, I am including the driver name for the vendor parameters. That enables validation with an admission webhook and providing setup parameters for different drivers, which can happen now when the CEL expression allows it

```yaml
requests:
- setup:
    vendor:
      driverName: cards.dra.example.com
      parameters: <parameters for devices A and B>
  devices:
  - selector: <selector for device A>
    setup:
      vendor:
        driverName: cards.dra.example.com
        parameters: <parameters for device A>
  - selector: selector for device B>
    setup:
      vendor:
        driverName: cards.dra.example.com
        parameters: <parameters for device B>
```

Reply:
Even better, we don't even need the top-level "requests".

See the update that I just pushed.